Custom serializers accessors

OlliTietavainenVaadin opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/10658)
<!-- Reviewable:end -->

OlliTietavainenVaadin committed:
add accessor methods for CUSTOM_SERIALIZERS in JsonCodec

OlliTietavainenVaadin committed:
Merge branch 'master' of https://github.com/vaadin/framework into custom-serializers-accessors

OlliTietavainenVaadin committed:
javadoc

Legioth commented:
*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L6GJzrbbT3eP9mILE5z:b-r7ry6t) (no related file):*
I'm slightly weary of simply exposing the current mechanism. `JSONSerializer` is not intended for general use and has not been reviewed as such (which can among other things be seen from the sloppily chosen name, should really be `JsonSerializer` to follow Java conventions).

Furthermore, there is no documentation and no code examples on how this functionality would be used in practice (i.e. how do you register a custom serializer, how is the client-side deserializing handled?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin commented:
Message from original requester of the change:
> We have written a custom serialization solution on the client side (i.e. TypeDataStore.get().setSerializerFactory(Query.class, new QuerySerializerFactory());) to optimize and limit the amount of information that is transmitted in the json message and we would like to have the corresponding custom deserialization capability on the receiving server side is the reason for the request. Does that answer the question?

elmot commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L6SGKzwDe9xcUfvbol2:boy5hoh) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

I'm slightly weary of simply exposing the current mechanism. `JSONSerializer` is not intended for general use and has not been reviewed as such (which can among other things be seen from the sloppily chosen name, should really be `JsonSerializer` to follow Java conventions).

Furthermore, there is no documentation and no code examples on how this functionality would be used in practice (i.e. how do you register a custom serializer, how is the client-side deserializing handled?)
</blockquote></details>

IMHO we can unlock the API here, despite of non-existing documentation. JSONSerializer has some javadocs, it has an example in open-source code. We have also RequestHandler, SessionInitListener and many other internal classes open and without good documentation coverage.

Yes, overwriting serializers might be a good way to shoot own leg, but is it our goal, to protect people from doing that at all costs?

Nevertheless, javadocs should be in places, and some warnings about legs and rounds should be there.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L6VWzxDu-6QvIfqVsB9:bb5937t) (no related file):*
<details><summary><i>Previously, elmot (Ilia Motornyi) wrote…</i></summary><blockquote>

IMHO we can unlock the API here, despite of non-existing documentation. JSONSerializer has some javadocs, it has an example in open-source code. We have also RequestHandler, SessionInitListener and many other internal classes open and without good documentation coverage.

Yes, overwriting serializers might be a good way to shoot own leg, but is it our goal, to protect people from doing that at all costs?

Nevertheless, javadocs should be in places, and some warnings about legs and rounds should be there.
</blockquote></details>

At the very least, I think the framework sources themselves should contain an end-to-end IT that uses matching server-side and client-side `JSONSerializer` implementation. Without this, we wouldn't even catch trivial regressions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

elmot commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L6qUn2n96tJOktHUZRB:bufn62) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

At the very least, I think the framework sources themselves should contain an end-to-end IT that uses matching server-side and client-side `JSONSerializer` implementation. Without this, we wouldn't even catch trivial regressions.
</blockquote></details>

Is that related to the pull request somehow?
Some API is just unlocked? whithout changing the behaviour
And looks like any of our IT tests are end-to-end tests for the serializer.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
Merge branch 'master' of https://github.com/vaadin/framework into custom-serializers-accessors

OlliTietavainenVaadin committed:
removed removeCustomSerializer method, renamed putCustomSerializer to addCustomSerializer, added sanity checks and JavaDocs

OlliTietavainenVaadin commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L8ba6dWAlWcxi3uXa87:bppaulg) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

I don't see anything that shows that the current behaviour is appropriate for usage from application code.

I don't see any test that would break if the `putCustomSerializer` method would accidentally be removed or its behaviour changed into a no-op.

Furthermore, the design of this functionality seem to be quite lacking (which would to some degree also have been uncovered by actually writing tests for it).
* Where should the application developer put their code that runs `putCustomSerializer`?
* What if trying to register multiple serializers for the same type?
* What if trying to register a serializer for some natively supported type?
* What if `putCustomSerializer?` is called concurrently from multiple threads?
* What if either method parameter is `null`?
* Why do we allow registering a `CustomSerializer<Foo>` for `String.class`? (I know it cannot be completely forbidden because of type erasure, but sensible use of type parameters would make it hard to cause problems by accident)
</blockquote></details>

Added some javadocs and sanity checks at least

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 456 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u7t8aEyMBYMd6sm4l:-L8ba47D2EpXKx_waj9Z:bubdpe5) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L456)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

`put` in the name is a strong reference to an implementation detail.
</blockquote></details>

Changed to add

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 457 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u9QvbEP1Folqrwqi5:-L8b_v6c85Jx6fQ4trdu:bcjid7e) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L457)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Why do we allow registering a `CustomSerializer<Foo>` for `String.class`?
</blockquote></details>

Throws an exception if trying to add a serializer for "java.lang" package class

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 458 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u90s39TyKSq-4QKdI:-L8b_u52AjY7mAWoAso0:b2m7ssf) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L458)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What about null values?
</blockquote></details>

Added null checks

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 462 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u93Tm9mEUOGwEOa0v:-L8b_siEB3DNco_EXgku:b58rzly) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L462)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

`CUSTOM_SERIALIZERS` sounds like an implementation detail.
</blockquote></details>

Removed method

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 468 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u9BLW9T8actSlDLc4:-L8b_pBkESYuVn8qwYeV:b-7hargg) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L468)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What's the purpose of this method? When would someone ever want to invoke it?
</blockquote></details>

Made it just for symmetry, but I guess it's not needed

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 469 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u99ae3rfXLa736iOF:-L8b_oAa-0nz_1YHo132:b58rzly) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L469)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

What about `null`
</blockquote></details>

Removed method

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

Legioth commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-L8fkxPr8617Cw7aSfE4:b-aros3i) (no related file):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Added some javadocs and sanity checks at least
</blockquote></details>

All but one of the bullet point questions from my previous comment still remain unanswered.

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 457 at r2](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L6u9QvbEP1Folqrwqi5:-L8fj7h22aynMvLOL8_L:beawcsm) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L457)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Throws an exception if trying to add a serializer for "java.lang" package class
</blockquote></details>

My point was that the wildcard type parameters (`<?>`) make it easy to accidentally do something analogous to `addCustomSerializer(Foo.class, new JsonSerializer<Bar> { ... })`.

It would be slightly weird, but I don't see why we would want to forbid adding a serializer for e.g. `java.lang.ref.WeakReference`.

---

*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 449 at r3](https://beta.reviewable.io/reviews/vaadin/framework/10658#-L8fkc2ODD7OLw0PfAWy:-L8fkc2ODD7OLw0PfAWz:b-p1vxhr) ([raw file](https://github.com/vaadin/framework/blob/ae897bfe22c175d682bab4a66f22385138718cc6/server/src/main/java/com/vaadin/server/JsonCodec.java#L449)):*
> ```Java
> 
>     /**
>      * Add a custom JSONSerializer for a specific Class. Friendly warning: it's
> ```

This warning is very unfriendly, since it only offers more questions and no answers to the reader.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vaadin/framework/10658)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin commented:
Closing

OlliTietavainenVaadin committed:
Merge branch 'master' into custom-serializers-accessors

OlliTietavainenVaadin committed:
Merge branch 'master' into custom-serializers-accessors

OlliTietavainenVaadin committed:
Merge branch 'master' of https://github.com/vaadin/framework into custom-serializers-accessors

OlliTietavainenVaadin committed:
refactored addCustomJsonSerializer to set, added test UI

OlliTietavainenVaadin committed:
Merge branch 'master' of https://github.com/vaadin/framework into custom-serializers-accessors

OlliTietavainenVaadin committed:
Merge branch 'custom-serializers-accessors' of https://github.com/OlliTietavainenVaadin/framework into custom-serializers-accessors

Legioth reviewed (changes requested):
> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 1 of 2 files reviewed, 6 unresolved discussions (waiting on @OlliTietavainenVaadin)

___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 451 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5c4b8FkLLGwHlz-f:-LWq5c4b8FkLLGwHlz-g:b-qvyp0r) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L451)):*
> ```Java
>      * Set a custom JSONSerializer for a specific Class. Existence of custom
>      * serializers is checked after basic types (Strings, Booleans, Numbers,
>      * Characters), Collections, Maps and Enums, so setting custom serializers
> ```

I'm wondering if it would make sense to also support custom serializers for enums?
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 453 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5Ye97l2bTJgPwzHf:-LWq5Ye97l2bTJgPwzHg:b2nh7mt) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L453)):*
> ```Java
>      * Characters), Collections, Maps and Enums, so setting custom serializers
>      * for these won't have any effect.
>      *
> ```

<p>
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LWq8niAEpdALd7BZZEF:b-7y0k3w) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
> ```Java
>      * for these won't have any effect.
>      *
>      * Warning: overriding existing custom serializers may lead into unexpected
> ```

This warning seems quite useless since someone calling this method has no way of knowing whether there are any custom serializers that they might accidentally overwrite.
___
*[uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java, line 57 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq7Xhc1TiDqOdDJo0f:-LWq7Xhc1TiDqOdDJo0g:bb7mwae) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java#L57)):*
> ```Java
>     @Override
>     protected void setup(VaadinRequest request) {
>         addSerializer();
> ```

Seems quite weird that a new global serializer is created and assigned every time a new UI is initialized. Furthermore, this can cause severe race conditions if multiple UI instances are created concurrently since they would all try to to update the same `HashMap` at the same time.


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
move enums to be parsed after custom serializers

OlliTietavainenVaadin committed:
move adding custom serializer to static block

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @Legioth)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LWze-mnBvkhiZgU1gmE:b9vl6to) (no related file):*
> Where should the application developer put their code that runs putCustomSerializer?
Static initializer or ServletContextListener or similar.
> What if trying to register multiple serializers for the same type?
Last will take precedence (name change from `add` to `set`should clarify
> What if trying to register a serializer for some natively supported type?
That will be ignored (javadoc comment points this out)
> What if putCustomSerializer? is called concurrently from multiple threads?
Probably bad things will happen, but it's not a very likely usage scenario.
> What if either method parameter is null?
If first parameter is null, it should throw (explained in javadoc). If second parameter is null, it should remove the serializer.

---
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 457 at r2](https://reviewable.io/reviews/vaadin/framework/10658#-L6u9QvbEP1Folqrwqi5:-LWzhDIXDGlqCdFQaCrq:b-w5t0sh) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L457)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

My point was that the wildcard type parameters (`<?>`) make it easy to accidentally do something analogous to `addCustomSerializer(Foo.class, new JsonSerializer<Bar> { ... })`.

It would be slightly weird, but I don't see why we would want to forbid adding a serializer for e.g. `java.lang.ref.WeakReference`.
</blockquote></details>

Removed the check. Do you mean adding a generic parameter and using that on the parameter classes?
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 451 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5c4b8FkLLGwHlz-f:-LWzdVHN-_H8vzGhsS_e:b-apao0j) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L451)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

I'm wondering if it would make sense to also support custom serializers for enums?
</blockquote></details>

Switched the evaluation order.
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 453 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5Ye97l2bTJgPwzHf:-LWzhZnS6FkDOdZybZp5:b-8sv620) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L453)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

<p>
</blockquote></details>

Empty comment?
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LWzdb2f3FJcMjg4so21:btf32t) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

This warning seems quite useless since someone calling this method has no way of knowing whether there are any custom serializers that they might accidentally overwrite.
</blockquote></details>

Added documentation
___
*[uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java, line 57 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq7Xhc1TiDqOdDJo0f:-LWzdjwh1SGCZxIJugS-:ban7xgl) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java#L57)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Seems quite weird that a new global serializer is created and assigned every time a new UI is initialized. Furthermore, this can cause severe race conditions if multiple UI instances are created concurrently since they would all try to to update the same `HashMap` at the same time.
</blockquote></details>

Moved to static initializer instead


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LX-8v0kAcEX0RWoIy9H:bagjl2j) (no related file):*
>> Where should the application developer put their code that runs putCustomSerializer?

> Static initializer or ServletContextListener or similar.

How will the user know that this is the way they're supposed to do it? Will they have any chance of realizing the mistake if they do it in some other way?

>> What if trying to register multiple serializers for the same type?

> Last will take precedence (name change from add to setshould clarify

Changing it to instead throw an exception would help detect situations where someone by accident forgets a `static` keyword for their initializer or something similar.

>> What if putCustomSerializer? is called concurrently from multiple threads?

> Probably bad things will happen, but it's not a very likely usage scenario.

The two first revisions of the example code in this PR indicates that it's a quite likely usage scenario.

---
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 457 at r2](https://reviewable.io/reviews/vaadin/framework/10658#-L6u9QvbEP1Folqrwqi5:-LWzxuiU8W1S0IrBNorb:b-cqdm9x) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L457)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Removed the check. Do you mean adding a generic parameter and using that on the parameter classes?
</blockquote></details>

Yes, something along the lines of `public static <T> void setCustomSerializer(Class<T> clazz, JSONSerializer<T> jsonSerializer)`.
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 453 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5Ye97l2bTJgPwzHf:-LWzyMBWCLB-tyVD5P2W:bhtitd3) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L453)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Empty comment?
</blockquote></details>

Silly reviewable that nowadays treats HTML as HTML. I tried to write `<p>` here.
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LWzyUck5tPPE7XjVJEa:bl869qu) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Added documentation
</blockquote></details>

Still doesn't help. With the current implementation, anyone using this method cannot do anything except hope for the best. The warning serves no purpose unless there's something that a developer can do to mitigate the risk.
___
*[uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java, line 57 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq7Xhc1TiDqOdDJo0f:-LX-8mmr82CfPpy0HZLF:b-qr1696) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java#L57)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Moved to static initializer instead
</blockquote></details>

Except that you forgot the `static` keyword in front of the initializer block.


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
throw an exception if multiple serializers are registered for class

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @Legioth)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LX-TCXX-KR-Mob36z1h:b4j9iy4) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

>> Where should the application developer put their code that runs putCustomSerializer?

> Static initializer or ServletContextListener or similar.

How will the user know that this is the way they're supposed to do it? Will they have any chance of realizing the mistake if they do it in some other way?

>> What if trying to register multiple serializers for the same type?

> Last will take precedence (name change from add to setshould clarify

Changing it to instead throw an exception would help detect situations where someone by accident forgets a `static` keyword for their initializer or something similar.

>> What if putCustomSerializer? is called concurrently from multiple threads?

> Probably bad things will happen, but it's not a very likely usage scenario.

The two first revisions of the example code in this PR indicates that it's a quite likely usage scenario.
</blockquote></details>

Throws an exception now if same key has been registered previously. I don't think there's any way to detect code is being called from a static context?

---
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 457 at r2](https://reviewable.io/reviews/vaadin/framework/10658#-L6u9QvbEP1Folqrwqi5:-LX-T6ua-M1RQQPo2RdH:bb6un1j) ([raw file](https://github.com/vaadin/framework/blob/1101de3f862ff3d0b526f1c705dd27d6c3eba122/server/src/main/java/com/vaadin/server/JsonCodec.java#L457)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Yes, something along the lines of `public static <T> void setCustomSerializer(Class<T> clazz, JSONSerializer<T> jsonSerializer)`.
</blockquote></details>

Done
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 453 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq5Ye97l2bTJgPwzHf:-LX-TMHM32FfqB5guNYz:bb6un1j) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L453)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Silly reviewable that nowadays treats HTML as HTML. I tried to write `<p>` here.
</blockquote></details>

Done
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LX-TPPt4D7AZvKf5set:b5d93dp) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Still doesn't help. With the current implementation, anyone using this method cannot do anything except hope for the best. The warning serves no purpose unless there's something that a developer can do to mitigate the risk.
</blockquote></details>

Now overriding should only happen by intentionally removing a previously-set serializer.
___
*[uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java, line 57 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq7Xhc1TiDqOdDJo0f:-LX-T8qy3aqogX6b34mF:bb6un1j) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/uitest/src/main/java/com/vaadin/tests/application/CustomJSONSerializer.java#L57)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Except that you forgot the `static` keyword in front of the initializer block.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LX4H0vE4xUt4beeJVYF:bkppmy) (no related file):*
>  I don't think there's any way to detect code is being called from a static context?

Not directly. What we could do, but I'm not sure if it's worthwhile, is to set a flag after e.g. `VaadinService.init` has run and then make the setter throw if the flag has been set. This is however slightly problematic because there can in theory be multiple `VaadinService` instances sharing the same `JsonCodec` class.

---
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LX4HoN08eyVPH-rhLU-:b-2z98vf) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Now overriding should only happen by intentionally removing a previously-set serializer.
</blockquote></details>

How does one remove a previously set serializer? I don't see any remove method.

Also, the warning might need some rephrasing.


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
updated javadocs

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LXOuxZo2oKeb6dKXI3h:b-ih0z1z) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

>  I don't think there's any way to detect code is being called from a static context?

Not directly. What we could do, but I'm not sure if it's worthwhile, is to set a flag after e.g. `VaadinService.init` has run and then make the setter throw if the flag has been set. This is however slightly problematic because there can in theory be multiple `VaadinService` instances sharing the same `JsonCodec` class.
</blockquote></details>

There's already the `initialized` flag in VaadinService that does exactly what you described, but I tried using it and it already failed in the uitest project, so maybe it's better to keep it as it is.

---
___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 454 at r5](https://reviewable.io/reviews/vaadin/framework/10658#-LWq8niAEpdALd7BZZEE:-LXOvPZp6R5pXQB00hCQ:bum33er) ([raw file](https://github.com/vaadin/framework/blob/8a7ed3732fc6467b90e23646f44b818230f0e2dc/server/src/main/java/com/vaadin/server/JsonCodec.java#L454)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

How does one remove a previously set serializer? I don't see any remove method.

Also, the warning might need some rephrasing.
</blockquote></details>

You can remove it by setting the serializer for a class to be null. I added javadoc for this. I also rephrased the warning a bit, but any improvements are welcome too.


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> Reviewed 2 of 3 files at r8, 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 3 unresolved discussions (waiting on @OlliTietavainenVaadin)

___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 455 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrr8NUA6_CE7cMDZ5I:-LXrr8NUA6_CE7cMDZ5J:b-iuakjv) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/server/src/main/java/com/vaadin/server/JsonCodec.java#L455)):*
> ```
> @code{null}
> ```

`{@code null}`.  Repeat below.
___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LXrreVW7zfDR2cKxlgX:bj0zi0b) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
> ```Java
> import elemental.json.JsonValue;
> 
> public class CustomJSONSerializerTest extends MultiBrowserTest {
> ```

Why is this a TestBench test that is run separately for each configured browser?


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
changed CustomJSONSerializerTest to a SingleBrowserTest

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 1 of 3 files reviewed, 3 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

___
*[server/src/main/java/com/vaadin/server/JsonCodec.java, line 455 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrr8NUA6_CE7cMDZ5I:-LXsmt2G89Bfqt40e8F2:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/server/src/main/java/com/vaadin/server/JsonCodec.java#L455)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

> ```
> @code{null}
> ```

`{@code null}`.  Repeat below.
</blockquote></details>

Done.
___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LXsmumt1-y9ukPjSgk4:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Why is this a TestBench test that is run separately for each configured browser?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> Reviewed 2 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LY22WUl881pmBCdi1J1:b-aswyqp) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Why is this a TestBench test at all, regardless of how many times it's run?


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LY5e4eoDCm6h_WrIOjh:b-4o1tho) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

Why is this a TestBench test at all, regardless of how many times it's run?
</blockquote></details>

Would unit test be better?


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LYQYL2SC51GCRraJBCW:bwbn8m4) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Would unit test be better?
</blockquote></details>

This is effectively a unit test already since it doesn't seem to use a browser at all.


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
moved CustomJSONSerializerTest to server/ and it's now not a browser test

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LYg9uyJ96Lv9jd0kbNW:bamjy1) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

This is effectively a unit test already since it doesn't seem to use a browser at all.
</blockquote></details>

Now it's a unit test.


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> Reviewed 1 of 1 files at r11.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LZ-ctCn0gzeoSqHa1dy:b-bgmpav) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Now it's a unit test.
</blockquote></details>

I guess this file should also be removed?


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
removed CustomJSONSerializerTest

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

___
*[uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java, line 16 at r9](https://reviewable.io/reviews/vaadin/framework/10658#-LXrreVW7zfDR2cKxlgW:-LZ0z2FU0qjYrQe-H854:b2nhswy) ([raw file](https://github.com/vaadin/framework/blob/a27ecffef492828ba1ae84c2445a98c028600636/uitest/src/test/java/com/vaadin/tests/application/CustomJSONSerializerTest.java#L16)):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

I guess this file should also be removed?
</blockquote></details>

Yes


<!-- Sent from Reviewable.io -->

Legioth reviewed (changes requested):
> Reviewed 1 of 1 files at r12.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-LZ3hXH_BJ4VQl5sXQmz:-LZ3hXH_BJ4VQl5sXQn-:b-ofpp21) (no related file):*
No further remarks from my side, but leaving final approval to someone from the actual development team. (Just dismiss me from this discussion)

---


<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin commented:
pinging @ZheSun88

ZheSun88 committed:
Merge branch 'master' into custom-serializers-accessors

OlliTietavainenVaadin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 2 unresolved discussions (waiting on @Legioth and @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-LZ3hXH_BJ4VQl5sXQmz:-LZIxGBN6UETuS59EaTd:b-6hsdgv) (no related file):*
<details><summary><i>Previously, Legioth (Leif Åstrand) wrote…</i></summary><blockquote>

No further remarks from my side, but leaving final approval to someone from the actual development team. (Just dismiss me from this discussion)
</blockquote></details>

@zhesun88 ok to you?

---


<!-- Sent from Reviewable.io -->

ZheSun88 committed:
Merge branch 'master' into custom-serializers-accessors

ZheSun88 reviewed (approved):
> Reviewed 1 of 3 files at r8, 1 of 2 files at r10, 1 of 1 files at r11, 1 of 1 files at r12.
> Dismissed @Legioth from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 1 unresolved discussion (waiting on @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-LZ3hXH_BJ4VQl5sXQmz:-LZJCCWs3lUwHO0Rwwad:b-r8uffn) (no related file):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

@zhesun88 ok to you?
</blockquote></details>

oh yes, now the branch-out-dated issue..  sorry, last time i triggered the build and it has been outdated again this time

---


<!-- Sent from Reviewable.io -->

ZheSun88 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/10658)* status: all files reviewed, 1 unresolved discussion (waiting on @OlliTietavainenVaadin)

*[a discussion](https://reviewable.io/reviews/vaadin/framework/10658#-L6GJzraFHaWLCahJo5R:-LZJCc-g1xAJ9D3n_WRK:bldj6e1) (no related file):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

There's already the `initialized` flag in VaadinService that does exactly what you described, but I tried using it and it already failed in the uitest project, so maybe it's better to keep it as it is.
</blockquote></details>

@ollitietavainenvaadin you are blocking yourself in this discussion i think

---


<!-- Sent from Reviewable.io -->
